### PR TITLE
[JENKINS-58780] Jenkins cancels the wrong build, when using QTF.cancel()

### PR DESCRIPTION
**Pull request by HedAurabesh**
When cancelling a build for a specific Future object, Jenkins currently does not care about actually cancelling the queue item that corresponds to the given future. Instead, it simply cancel the first build for the same task that it can find.

This can be easily replicated by starting two jobs with different parameters via a Groovy script without having any build hosts online (so that they get queued).

The script should hold onto the returned QueueTaskFuture instances and then issue a cancellation on only one of them.

Current Jenkins behaviour is to always cancel the first of the tasks, no matter which of the two future objects were called.

This does not just affect programmatic calling of builds, but can also happen in downstream-triggered tasks and other build starting methods that rely on a cancellation via the QueueTaskFuture instance being reliable.

A new Unittest is being provided in this commit for this scenario.

Signed-off-by: Martin H Schroeder <martin.h.schroeder@intel.com>

See [JENKINS-58780](https://issues.jenkins-ci.org/browse/JENKINS-58780).

## Changelog entry

* Bug: Prevent cancelling wrong tasks in the Jenkins queue when the `QueueTaskFuture#cancel()` API is called

### CODE REVIEW

File: core/src/main/java/hudson/model/Queue.java [hunk 1]
```diff
@@ -1,9 +1,12 @@
 /*
  * The MIT License
- *
+ * 
+ * Copyright (c) 2018-2019 Intel Corporation
+ * Copyright (c) 2015-2017, Intel Deutschland GmbH
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java]: Just add "and other contributors to the original title?"

Comment (HedAurabesh) [reply core/src/main/java/hudson/model/Queue.java]: Due to legal requirements, Intel policy does not permit us to join the previous business entities under which Intel operated here in German.

This leads us to having to include the following copyrights depending on the year it was issued:

Copyright (c) 2018-2019 Intel Corporation
Copyright (c) 2015-2017, Intel Deutschland GmbH
Copyright (c) 2011-2014, Intel Mobile Communications GmbH

And yes, the lack of a comma in the 2018/2019 version is intentional. Isn't copyright law fun? :)

Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/Queue.java]: As an ex-Intel employee, I know that feeling

What if...

* The original copyright is updated to `2004-2019` and to reference "all contributors" after it
* Intel copyrights are listed below the generic ones

Comment (HedAurabesh) [reply core/src/main/java/hudson/model/Queue.java]: Hi Oleg.
I got in touch with our open source copyright legal contact and they got us a waiver that we can condense the copyrights, since we didn't actually release anything before our business entity had already been completely absorbed.

I will change this pull request and 4138 (that I uploaded at the same time) accordingly.

Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/Queue.java]: thanks!

**Review by varyvol — APPROVED** [review]
I don't have a strong knowledge of this part of the code but seems good.

**Review by oleg-nenashev — APPROVED** [review]
I think this change should work well.
CC @jglick @stephenc just in case, more reviews would be helpful here

File: core/src/main/java/hudson/model/Queue.java [hunk 2]
```diff
@@ -1,9 +1,10 @@
 /*
  * The MIT License
- *
+ *
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: Please revert unrelated changes so that a diff is minimal before submitting a PR. (Makes reviews easier, reduces changes of merge conflicts, avoids noise in `git blame`, etc.)

File: core/src/main/java/hudson/model/Queue.java [hunk 3]
```diff
@@ -675,6 +677,14 @@ public WaitingItem schedule(Task p, int quietPeriod, List<Action> actions) {
         }
     }
 
+    private void updateSnapshot() {
+        Snapshot revised = new Snapshot(waitingList, blockedProjects, buildables, pendings);
+        if (LOGGER.isLoggable(Level.FINEST)) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: BTW you can also use the overload taking a `Supplier` lambda, which is more convenient.

File: core/src/main/java/hudson/model/Queue.java [hunk 4]
```diff
@@ -713,7 +723,16 @@ public boolean add(Task p, int quietPeriod, Action... actions) {
     }
 
     /**
-     * Cancels the item in the queue. If the item is scheduled more than once, cancels the first occurrence.
+     * Cancels the item in the queue. If the item is scheduled more than once,
+     * cancels the <b>first</b> occurrence.
+     * <p>
+     * Be aware, that the same task can be multiple times in the Queue, e.g. if
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]:
```suggestion
     * Be aware, that the same task can be present multiple times in the Queue, e.g. if
```

File: core/src/main/java/hudson/model/Queue.java [hunk 5]
```diff
@@ -713,7 +723,16 @@ public boolean add(Task p, int quietPeriod, Action... actions) {
...
+     * parameters are set on the project. In that case, it is Russian Roulette,
+     * whether or not you cancel the correct task.
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: So does it make sense to deprecate this overload, or are there cases (at least from searching usages within this repo) where it is impractical to use the other overloads?

There are several new methods being introduced, but zero examples of them being used successfully, either here or in some downstream PR using an incremental deployment. Any proposed new API must have at least one matching proposed call (outside of test code) which proves that the API is useful. If none can be found, then the method should be deleted. And if no proposes usage can be found outside of this repository (in a plugin), then it should be

```java
@Restricted(NoExternalUse.class)
```

unless and until one is found.

File: core/src/main/java/hudson/model/Queue.java (hunk 3 above)
Comment (jglick) [reply core/src/main/java/hudson/model/Queue.java]: Ah but it looks like this method was just moved around? If so, better to return it to its original position, to minimize the diff.

File: core/src/main/java/hudson/model/Queue.java [hunk 6]
```diff
@@ -2937,13 +2997,34 @@ public void put(Task task, T item) {
...
+        
+        /**
+         * Marks the {@link Item} associated with the given Future as cancelled.
+         * <p>
+         * The {@link Queue} itself must ensure that the write-lock on
+         * {@link Queue#qLock} has been acquired before calling this method.
+         */
+        public T cancel(FutureImpl f) {
+            T x = get(f);
+            if(x!=null) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]:
```suggestion
            if (x != null) {
```

File: core/src/main/java/hudson/model/queue/FutureImpl.java [hunk 7]
```diff
@@ -70,17 +78,34 @@ public final Executable waitForStart() throws InterruptedException, ExecutionExc
...
-                    if(mayInterruptIfRunning)
-                        for (Executor e : executors)
-                            e.interrupt();
-                    return mayInterruptIfRunning;
+        final boolean mayInterrupt = mayInterruptIfRunning;
+        final FutureImpl self = this;
+        
+        //Cancel while holding an exclusive lock on the queue and this future
+        //Note, this is the same order of locking as before, when the queue was also synchronized
+        Callable<Boolean> worker = new Callable<Boolean> () {
```
Comment (jglick) [inline core/src/main/java/hudson/model/queue/FutureImpl.java]: Could convert to lambda for legibility. (Then also `self` is unnecessary, as `this` would be a `FutureImpl`.)

File: test/src/test/java/hudson/model/QueueTest.java [hunk 8]
```diff
@@ -1187,4 +1193,111 @@ public void run() {
...
+        
+        //Ensure that only a single executor is available
+        r.jenkins.setNumExecutors(1);
+        r.jenkins.reload();
+        
+        //Schedule ONE job that will block the one executor
+        //This is done, so the other jobs will be buildable, not blocked
+        q.schedule2(t, 0);
+        //Wait, for it to be scheduled
+        Thread.sleep(1000);
```
Comment (jglick) [inline test/src/test/java/hudson/model/QueueTest.java]: Can the test wait for a specific condition to hold, rather than just sleeping and hoping for the best? This looks like it will be flaky.

File: core/src/main/java/hudson/model/queue/FutureImpl.java [hunk 9]
```diff
@@ -60,6 +64,10 @@ public FutureImpl(Task task) {
         this.task = task;
     }
 
+    public String toString() {
```
Comment (jglick) [inline core/src/main/java/hudson/model/queue/FutureImpl.java]:
```suggestion
    @Override
    public String toString() {
```

**Review by jglick — CHANGES_REQUESTED** [review]
The original method does indeed sound unsafe, but so far it is unclear how this PR is going to help.

**Comment by HedAurabesh** [comment]
Hi everyone.

Thanks for the feedback! I will incorporate your suggestions and upload a new commit soon -- or reply to the reviews where necessary.

File: core/src/main/java/hudson/model/queue/FutureImpl.java [hunk 10]
```diff
@@ -70,17 +78,34 @@ public final Executable waitForStart() throws InterruptedException, ExecutionExc
...
-                            e.interrupt();
-                    return mayInterruptIfRunning;
+        final boolean mayInterrupt = mayInterruptIfRunning;
+        final FutureImpl self = this;
+        
+        //Cancel while holding an exclusive lock on the queue and this future
+        //Note, this is the same order of locking as before, when the queue was also synchronized
+        Callable<Boolean> worker = new Callable<Boolean> () {
+            public Boolean call() {
+                Queue q = Jenkins.getInstance().getQueue();
```
Comment (res0nance) [inline core/src/main/java/hudson/model/queue/FutureImpl.java]:
```suggestion
                Queue q = Jenkins.get().getQueue();
```
Use non deprecated `get()` instead of `getInstance()`

**Comment by oleg-nenashev** [comment]
@HedAurabesh hi, do you still plan to work on it?

**Comment by HedAurabesh** [comment]
> @HedAurabesh hi, do you still plan to work on it?

Hi Oleg. Sorry for the late reply.

Yes, I do plan to work on it. Right now our part of the company is just going through a lot of upheavels, which makes working on some things much less prioritized. But as soon as that calms down, I will resume work on this pull request.

**Comment by varyvol** [comment]
@HedAurabesh given you didn't have the availability to work on this for some time now, I'm proposing it to be closed. Please feel free to reach us if you think otherwise.
